sdk/typescript/src/signers/txn-data-serializers/call-arg-serializer.ts

```diff
@@ -0,0 +1,169 @@
+        async (param, i) => await this.newCallArg(param, txn.arguments[i])
+      )
+    );
+  }
+
+  async newObjectArg(objectId: string): Promise<ObjectArg> {
+    const object = await this.provider.getObject(objectId);
+    if (isSharedObject(object)) {
+      return { Shared: objectId };
+    }
+
+    return { ImmOrOwned: getObjectReference(object)! };
+  }
+
+  private async newCallArg(
+    expectedType: SuiMoveNormalizedType,
+    argVal: SuiJsonValue
+  ): Promise<CallArg> {
+    const structVal = extractStructTag(expectedType);
+    if (structVal != null) {
+      if (typeof argVal !== 'string') {
+        throw new Error(
+          `${MOVE_CALL_SER_ERROR} expect the argument to be an object id string, got ${argVal}`
+        );
+      }
+      return { Object: await this.newObjectArg(argVal) };
+    }
+
+    let serType = this.getPureSerializationType(expectedType, argVal);
+
+    return {
+      Pure: bcs.ser(serType, argVal).toBytes(),
+    };
+  }
+
+  /**
+   *
+   * @param argVal used to do additional data validation to make sure the argVal
+   * matches the normalized Move types. If `argVal === undefined`, the data validation
+   * will be skipped. This is useful in the case where `normalizedType` is a vector<T>
+   * and `argVal` is an empty array, the data validation for the inner types will be skipped.
+   */
+  private getPureSerializationType(
+    normalizedType: SuiMoveNormalizedType,
+    argVal: SuiJsonValue | undefined
+  ): string {
+    const allowedTypes = ['Address', 'Bool', 'U8', 'U32', 'U64', 'U128'];
+    if (
+      typeof normalizedType === 'string' &&
+      allowedTypes.includes(normalizedType)
+    ) {
+      if (normalizedType in ['U8', 'U32', 'U64', 'U128']) {
+        this.checkArgVal(isTypeFunc('number'), argVal, 'number');
+      } else if (normalizedType === 'Bool') {
+        this.checkArgVal(isTypeFunc('boolean'), argVal, 'boolean');
+      } else if (normalizedType === 'Address') {
+        this.checkArgVal(
+          (t: any) => typeof t === 'string' && isValidSuiAddress(t),
+          argVal,
+          'valid SUI address'
+        );
+      }
+      return normalizedType.toLowerCase();
+    } else if (typeof normalizedType === 'string') {
+      throw new Error(
+        `${MOVE_CALL_SER_ERROR} unknown pure normalized type ${normalizedType}`
+      );
+    }
+
+    if ('Vector' in normalizedType) {
+      if (typeof argVal === 'string' && normalizedType.Vector === 'U8') {
+        return 'string';
+      }
+
+      if (!Array.isArray(argVal)) {
+        throw new Error(
+          `Expect ${argVal} to be a array, received ${typeof argVal}`
+        );
+      }
+      const innerType = this.getPureSerializationType(
+        normalizedType.Vector,
+        // undefined when argVal is empty
+        argVal[0]
+      );
+      const res = `vector<${innerType}>`;
+      // TODO: can we get rid of this call and make it happen automatically?
+      bcs.registerVectorType(res, innerType);
```

Comment:
Is there any chance the registered type conflicts between multiple packages? If so seems like we'd want a way to not have these registered into a global scope.

Reply:
Good point, will look into that